security.go

```diff
@@ -1,47 +1,91 @@
-	if header["X-Slack-Signature"][0] == "" || header["X-Slack-Request-Timestamp"][0] == "" {
-		return SecretsVerifier{}, errors.New("Headers are empty, cannot create SecretsVerifier")
-	}
-
-	hash := hmac.New(sha256.New, []byte(signingSecret))
-	hash.Write([]byte(fmt.Sprintf("v0:%s:", header["X-Slack-Request-Timestamp"][0])))
-	return SecretsVerifier{
-		slackSig:  header["X-Slack-Signature"][0],
-		timeStamp: header["X-Slack-Request-Timestamp"][0],
-		hmac:      hash,
-	}, nil
-}
-
-func (v *SecretsVerifier) Write(body []byte) (n int, err error) {
-	return v.hmac.Write(body)
-}
-
-// Ensure compares the signature sent from Slack with the actual computed hash to judge validity
-func (v SecretsVerifier) Ensure() error {
-	computed := "v0=" + string(hex.EncodeToString(v.hmac.Sum(nil)))
-	if computed == v.slackSig {
-		return nil
-	}
-
-	return fmt.Errorf("Expected signing signature: %s, but computed: %s", v.slackSig, computed)
-}
+package slack
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"hash"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+// Signature headers
+const (
+	hSignature = "X-Slack-Signature"
+	hTimestamp = "X-Slack-Request-Timestamp"
+)
+
+// SecretsVerifier contains the information needed to verify that the request comes from Slack
+type SecretsVerifier struct {
+	signature string
+	hmac      hash.Hash
+}
+
+func unsafeSignatureVerifier(header http.Header, secret string) (_ SecretsVerifier, err error) {
+	signature := header.Get(hSignature)
+	stimestamp := header.Get(hTimestamp)
+
+	if signature == "" || stimestamp == "" {
+		return SecretsVerifier{}, errors.New("missing headers")
+	}
+
+	hash := hmac.New(sha256.New, []byte(secret))
+	hash.Write([]byte(fmt.Sprintf("v0:%s:", stimestamp)))
+
+	return SecretsVerifier{
+		signature: signature,
+		hmac:      hash,
+	}, nil
+}
+
+// NewSecretsVerifier returns a SecretsVerifier object in exchange for an http.Header object and signing secret
+func NewSecretsVerifier(header http.Header, secret string) (sv SecretsVerifier, err error) {
+	var (
+		timestamp int64
+	)
+
+	stimestamp := header.Get(hTimestamp)
+
+	if sv, err = unsafeSignatureVerifier(header, secret); err != nil {
+		return sv, err
+	}
+
+	if timestamp, err = strconv.ParseInt(stimestamp, 10, 64); err != nil {
+		return SecretsVerifier{}, err
+	}
+
+	diff := absDuration(time.Now().Sub(time.Unix(timestamp, 0)))
+	if diff > 5*time.Minute {
+		return sv, fmt.Errorf("timestamp is too old")
+	}
+
+	return sv, err
+}
+
+func (v *SecretsVerifier) Write(body []byte) (n int, err error) {
+	return v.hmac.Write(body)
+}
+
+// Ensure compares the signature sent from Slack with the actual computed hash to judge validity
+func (v SecretsVerifier) Ensure() error {
+	computed := "v0=" + string(hex.EncodeToString(v.hmac.Sum(nil)))
+
+	// use hmac.Equal prevent leaking timing information.
+	if hmac.Equal([]byte(computed), []byte(v.signature)) {
```

Comment:
This seems overly-complicated. It computes the HMAC (`[]bytes`), hex encodes that to a `string`, then converts that back to `[]bytes` (along with the header signature) for comparison.

IMO it's more elegant to decode the header signature (removing the `"v0="` prefix) with [`hex.DecodeString()`](https://golang.org/pkg/encoding/hex/#DecodeString) and pass that and the computed HMAC directly to `hmac.Equal()`.

Reply:
haha good catch wasn't paying attention to the conversions.